content/advent-2016/sea-of-open-source.md

```diff
@@ -0,0 +1,165 @@
+versus writing your own?
+
+Let that sink in for a bit and now think of when you are learning a language
+and you don't really know what is considered idiomatic or not.
+
+Those are hard questions. They become even harder if you have to use these
+third-party packages to build applications that are going to be used into
+production code.
+
+Usually you have a checklist of items that you'd like a package or library
+to meet before you can use it. My list looks like this (not in the order of 
+importance):
+
+- is it used by others?
+- how does the documentation look like?
+- does it do what I need it to do (and preferably nothing more)?
+- is the author maintaining the package? How are issues / PRs addressed? Are
+there any contributors to the project?
+- license
+- what's the code quality like? Does it have tests?
+
+If you wonder why code quality is the last one on the list the answer is
+simple: because I can assert the first five in just a few minutes while
+reading the code can take a bit longer.
+
+So how do you actually find packages?
+
+For Go we have [godoc.org](https://godoc.org/) and [go-search.org](http://go-search.org/) which allow you
+to search for Go packages. With over 150.000 projects and packages indexed
+by them, finding a "good" package could quickly become a problem.
+
+Fortunately, both sort results by popularity (do not confuse with Github stars,
+we'll come back to this a bit later). My default search tool is GoDoc but you
+can also use GoSearch as GoDoc references it when you search for a package that
+is not found in its index.
+
+Popularity for Go packages means how often they are imported which will already
+answer the first question.
+
+Then I can quickly have a look at the documentation as well. And here is were
+one of the Go's lesser known features shines. The documentation of a package
+can be [tested](https://blog.golang.org/examples) so if I see that then I can
+have the guarantee that the project is already in a good shape.
+
+Having briefly checked the documentation, I can quickly go to the place where
+it's hosted and see what's the situation looking like for the other factors.
+
+Remember I've said that you shouldn't confuse popularity with number of Github
+stars? Well this is the time to ignore the Github stars as again. I'll stick
+with Github for now since it's more or less the de facto place for source code
+hosting.
+
+Back to our package, lets say that we are looking at a package which has a few
+contributors (that's a nice touch) and a few open issues and maybe a few PRs.
+All those things could indicate that in fact people are contributing to the
+package.
+
+The first thing I do is to look at the readme and see what's in there.
+Important information that I can quickly see is:
+- does the package have any badge for a CI server?
+- does the package have instructions for how to install / use?
+- are there any examples there that might be relevant? Or maybe an FAQ / link
+to a wiki?
```

Comment:
I also like looking at the "Latest commit" date to see if it's been updated in the past year 👍